src/main/java/com/yelp/nrtsearch/server/luceneserver/warming/ReservoirSampler.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2020 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.warming;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A simple implementation of <a href="https://en.wikipedia.org/wiki/Reservoir_sampling">reservoir
+ * sampling</a> used for having a set of warming queries.
+ */
+public class ReservoirSampler {
+
+  private final long maxQueries;
+  private final Map<String, AtomicLong> numQueries;
+
+  /** @param maxQueries Maximum number of warming queries. */
+  public ReservoirSampler(long maxQueries) {
+    this.maxQueries = maxQueries;
+    this.numQueries = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Decide if a query should be included for warming or not.
+   *
+   * @param indexName Index name for the query
+   * @return a {@link SampleResult}. If the query must be included, {@link SampleResult#isSample()}
+   *     is {@code true} and {@link SampleResult#getReplace()} points to the index of query in the
+   *     list to replace current query with. If the query must not be included then {@code isSample}
+   *     returns {@code false}.
+   */
+  public SampleResult sample(String indexName) {
+    long numQueriesForIndex =
+        numQueries.computeIfAbsent(indexName, s -> new AtomicLong(0)).getAndIncrement();
+    if (numQueriesForIndex < maxQueries) {
+      return new SampleResult(true, (int) numQueriesForIndex);
+    }
+    int replace = (int) ThreadLocalRandom.current().nextLong(numQueriesForIndex);
+    if (replace < maxQueries) {
+      return new SampleResult(true, replace);
+    }
+    return new SampleResult(false, 0);
```

Comment:
Since the SampleResult when not sampling is always the same, it could be a static final instance.